modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,304 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include "dlsobackend.hh"
+#include <dlfcn.h>
+
+static const char *kBackendId = "[DlsoBackend]";
+
+/**
+ * Standard ctor and dtor
+ */
+DlsoBackend::DlsoBackend(const std::string &suffix)
+{
+  setArgPrefix("dlso"+suffix);
+
+  std::string libpath = getArg("path");
+  std::string args = getArg("args");
+
+  this->d_dnssec = mustDo("dnssec");
+  this->in_query = false;
+
+  this->dlhandle = dlopen(libpath.c_str(), RTLD_LAZY);
+
+  if (!this->dlhandle) {
+      throw PDNSException("Unable to load library: " + libpath);
+  }
+
+  this->api.new_ = (void * (*)(bool, const char *)) dlsym(this->dlhandle, "pdns_dlso_new");
+  this->api.free_ = (void (*)(void *)) dlsym(this->dlhandle, "pdns_dlso_free");
```

Comment:
No check against NULL is done, which means if one makes a plugin that does not define pdns_dlso_free will result in a segmentation fault